common/src/main/java/com/github/khanshoaib3/minecraft_access/features/inventory_controls/GroupGenerator.java

```diff
@@ -467,8 +466,51 @@ else if (screen.getHandler() instanceof HopperScreenHandler)
             foundGroups.add(itemOutputGroup);
         }
 
-        if (unknownGroup.slotItems.size() > 0) {
-            foundGroups.add(unknownGroup);
+        // Adjacency is used here instead of slot.inventory because some mods have non-adjacent slots in the same
+        // inventory and adjacent slots in different inventories
+        List<List<SlotItem>> unknownGroups = new ArrayList<>(unknownSlots.size());
+        for (SlotItem slot : unknownSlots) {
+            unknownGroups.stream()
+                    .filter(group ->
+                            group.stream()
+                                    .anyMatch(groupSlot ->
+                                            groupSlot.x == slot.x && (groupSlot.y == slot.y + 18 || groupSlot.y == slot.y - 18)
+                                            || groupSlot.y == slot.y && (groupSlot.x == slot.x + 18 || groupSlot.x == slot.x - 18)
+                                    )
+                    )
```

Comment:
Can you extract these inner logic into methods?
Like:
 unknownGroups.stream()
                    .filter(GroupGenerator::<new method 1>)
                    .findFirst()
                    .or(GroupGenerator::<new method 2>)
                    .get()
                    .add(slot);

BTW I'm confused what this stream operation tries to do...

Reply:
The outer stream operation looks for an unknown group which already has adjacent slots in it using the inner stream which checks if any slots in the group are adjacent, if it doesn't find any it creates a new group and returns that instead.

Reply:
Upon further inspection it doesn't look like it's practical to extract those to methods because they reference stuff in the outer scope so I wouldn't be able to use method references.